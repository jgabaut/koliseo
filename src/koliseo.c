#include "koliseo.h"
int KOLISEO_DEBUG = 0;
FILE* KOLISEO_DEBUG_FP = NULL;

/**
 * Returns the constant string representing current version for Koliseo.
 * @return A constant string in MAJOR-MINOR-PATCH format for current Koliseo version.
 */
const char* string_koliseo_version(void) {
	return KOLISEO_API_VERSION_STRING;
}

/**
 * Returns the current offset (position of pointer bumper) for the passed Koliseo.
 * @param kls The Koliseo at hand.
 * @return A ptrdiff_t value for current position.
 */
ptrdiff_t kls_get_pos(Koliseo* kls) {
	return kls->offset;
}

/**
 * When KOLISEO_DEBUG is 1, logs a message to the defined KOLISEO_DEBUG_FP.
 * @param tag Tag for a message.
 * @param msg The actual message.
 */
void kls_log(const char* tag, const char* msg) {
	if (KOLISEO_DEBUG == 1) {
		if (KOLISEO_DEBUG_FP == NULL) {
			fprintf(stderr,"[KLS]    kls_log(): Failed opening KOLISEO_DEBUG_FP to print logs.\n");
		} else {
			time_t now = time(0);
			struct tm *mytime = localtime(&now);
			char timeheader[500];
			if ( strftime(timeheader, sizeof timeheader, "%X", mytime) )
			{
				fprintf(KOLISEO_DEBUG_FP,"[%-10.10s] [%s] [%s]\n", tag, timeheader, msg);
			}
		}
	}
}


/**
 * Takes a ptrdiff_t size and allocates the backing memory for a Koliseo. Sets the fields with appropriate values if memory allocation was successful, goes to abort() otherwise.
 * @param size The size for Koliseo data field.
 * @return A pointer to the initialised Koliseo struct.
 * @see Koliseo
 * @see Koliseo_Temp
 * @see kls_temp_start()
 * @see kls_temp_end()
 */
Koliseo* kls_new(ptrdiff_t size) {
	assert(size >= (ptrdiff_t)sizeof(Koliseo));
	void *p = malloc(size);
	if (p) {
		char msg[500];
		sprintf(msg,"Allocated (%li) for new KLS.",size);
		kls_log("KLS",msg);
		Koliseo* kls = p;
		kls->data = p;
		kls->size = size;
		kls->offset = sizeof(*kls);
		kls->prev_offset = kls->offset;
	} else {
		fprintf(stderr,"[KLS] Failed kls_new() call.\n");
		abort();
	}
	if (KOLISEO_DEBUG == 1) {
		print_kls_2file(KOLISEO_DEBUG_FP,p);
	}
	return p;
}

/**
 * Takes a Koliseo pointer, and ptrdiff_t values for size, align and count. Tries popping the specified amount of memory from the Koliseo data field, marking it as free (as far as Koliseo is concerned), or goes to abort() if the operation fails.
 * @param kls The Koliseo at hand.
 * @param size The size for data to pop.
 * @param align The alignment for data to pop.
 * @param count The multiplicative quantity to scale data size to pop for.
 * @return A void pointer to the start of memory just popped from the Koliseo.
 */
void* kls_pop(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count) {
	ptrdiff_t padding = -kls->offset & (align -1);
	if (count > PTRDIFF_MAX/size || (kls->size + kls->offset) < (size*count)) {
		fprintf(stderr,"[KLS] Failed kls_pop() call.\n");
		abort();
	}
	char* p = kls->data + kls->offset - padding - size*count;
	kls->prev_offset = kls->offset;
	kls->offset -= padding + size*count;
	char msg[500];
	sprintf(msg,"Popped (%li) for KLS.",size);
	kls_log("KLS",msg);
	if (KOLISEO_DEBUG == 1) {
		print_kls_2file(KOLISEO_DEBUG_FP,kls);
	}
	return p;
}

/**
 * Takes a Koliseo pointer, and ptrdiff_t values for size, align and count. Tries pushing the specified amount of memory to the Koliseo data field, or goes to abort() if the operation fails.
 * Notably, it does NOT zero the memory region.
 * @param kls The Koliseo at hand.
 * @param size The size for data to push.
 * @param align The alignment for data to push.
 * @param count The multiplicative quantity to scale data size to push for.
 * @return A void pointer to the start of memory just pushed to the Koliseo.
 */
void* kls_push(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count) {
	ptrdiff_t available = kls->size - kls->offset;
	ptrdiff_t padding = -kls->offset & (align -1);
	if (count > PTRDIFF_MAX/size || available-padding < size*count) {
		if (count > PTRDIFF_MAX/size) {
			fprintf(stderr, "[KSL]  count [%li] was bigger than PTRDIFF_MAX/size [%li].\n", count, PTRDIFF_MAX/size);
		} else {
			fprintf(stderr, "[KLS]  Out of memory. size*count [%li] was bigger than available-padding [%li].\n", size*count, available-padding);
		}
		fprintf(stderr,"[KLS] Failed kls_push() call.\n");
		abort();
	}
	char* p = kls->data + kls->offset + padding;
	kls->prev_offset = kls->offset;
	kls->offset += padding + size*count;
	char msg[500];
	sprintf(msg,"Pushed (%li) for KLS.",size);
	kls_log("KLS",msg);
	if (KOLISEO_DEBUG == 1) {
		print_kls_2file(KOLISEO_DEBUG_FP,kls);
	}
	return p;
}

/**
 * Takes a Koliseo pointer, and ptrdiff_t values for size, align and count. Tries pushing the specified amount of memory to the Koliseo data field, or goes to abort() if the operation fails.
 * Notably, it zeroes the memory region.
 * @param kls The Koliseo at hand.
 * @param size The size for data to push.
 * @param align The alignment for data to push.
 * @param count The multiplicative quantity to scale data size to push for.
 * @return A void pointer to the start of memory just pushed to the Koliseo.
 */
void* kls_push_zero(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count) {
	ptrdiff_t available = kls->size - kls->offset;
	ptrdiff_t padding = -kls->offset & (align -1);
	if (count > PTRDIFF_MAX/size || (available - padding) < (size*count)) {
		if (count > PTRDIFF_MAX/size) {
			fprintf(stderr, "[KSL]  count [%li] was bigger than PTRDIFF_MAX/size [%li].\n", count, PTRDIFF_MAX/size);
		} else {
			fprintf(stderr, "[KLS]  Out of memory. size*count [%li] was bigger than available-padding [%li].\n", size*count, available-padding);
		}
		fprintf(stderr,"[KLS] Failed kls_push_zero() call.\n");
		abort();
		//return 0;
	}
	char* p = kls->data + kls->offset + padding;
	//Zero new area
	memset(p, 0, size*count);
	kls->prev_offset = kls->offset;
	kls->offset += padding + size*count;
	char msg[500];
	sprintf(msg,"Pushed zeroes, size (%li) for KLS.",size);
	kls_log("KLS",msg);
	if (KOLISEO_DEBUG == 1) {
		print_kls_2file(KOLISEO_DEBUG_FP,kls);
	}
	return p;
}

/**
 * Prints header fields from the passed Koliseo pointer, to the passed FILE pointer.
 * @param kls The Koliseo at hand.
 */
void print_kls_2file(FILE* fp, Koliseo* kls) {
	if (fp == NULL) {
		fprintf(stderr,"print_kls_2file():  fp was NULL.\n");
		exit(EXIT_FAILURE);
	}
	if (kls == NULL) {
		fprintf(fp,"[KLS] kls was NULL.");
	} else {
		fprintf(fp,"\n[KLS] Size: [%li]\n", kls->size);
		fprintf(fp,"[KLS] Offset: [%li]\n", kls->offset);
		fprintf(fp,"[KLS] Prev_Offset: [%li]\n\n", kls->prev_offset);
	}
}

/**
 * Prints header fields from the passed Koliseo pointer, to stderr.
 * @param kls The Koliseo at hand.
 */
void print_dbg_kls(Koliseo* kls) {
  print_kls_2file(stderr,kls);
}

/**
 * Resets the offset field for the passed Koliseo pointer.
 * Notably, it sets the prev_offset field to the previous offset, thus remembering where last allocation was before the clear.
 * @param kls The Koliseo at hand.
 */
void kls_clear(Koliseo* kls) {
	//Reset pointer
	kls->prev_offset = kls->offset;
	kls->offset = sizeof(*kls);
	char msg[500];
	sprintf(msg,"Cleared offsets for KLS.");
	kls_log("KLS",msg);
}

/**
 * Calls kls_clear() on the passed Koliseo pointer and the frees the actual Koliseo.
 * @param kls The Koliseo at hand.
 * @see kls_clear()
 */
void kls_free(Koliseo* kls) {
	kls_clear(kls);
	free(kls);
	char msg[500];
	sprintf(msg,"Freed KLS.");
	kls_log("KLS",msg);
}

/**
 * Starts a new savestate for the passed Koliseo pointer, by initialising a Koliseo_temp and returning it.
 * Notably, you should not use the original while using the copy.
 * @param kls The Koliseo at hand.
 * @return A Koliseo_Temp struct.
 * @see Koliseo_Temp
 */
Koliseo_Temp kls_temp_start(Koliseo* kls) {
	Koliseo_Temp tmp;
	tmp.kls = kls;
	tmp.prev_offset = kls->prev_offset;
	tmp.offset = kls->offset;
	char msg[500];
	sprintf(msg,"Prepared new Temp KLS.");
	kls_log("KLS",msg);
	return tmp;
}

/**
 * Ends a new savestate for the passed Koliseo pointer, by initialising a Koliseo_temp and returning it.
 * Notably, you should not use the original while using the copy.
 * @param kls The Koliseo at hand.
 */
void kls_temp_end(Koliseo_Temp tmp_kls) {
	tmp_kls.kls->prev_offset = tmp_kls.prev_offset;
	tmp_kls.kls->offset = tmp_kls.offset;
	char msg[500];
	sprintf(msg,"Ended Temp KLS.");
	kls_log("KLS",msg);
}
