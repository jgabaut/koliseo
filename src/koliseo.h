#ifndef KOLISEO_H_
#define KOLISEO_H_
#include <stdio.h>
#include <stdint.h>
#include <stddef.h>
#include <stdlib.h>
#include <stdarg.h>
#include <assert.h>
#include <string.h>
#include <time.h>


#define KLS_MAJOR 0 /**< Represents current major release.*/
#define KLS_MINOR 2 /**< Represents current minor release.*/
#define KLS_PATCH 0 /**< Represents current patch release.*/

/**
 * Global variable for debug flag.
 */
extern int KOLISEO_DEBUG;
/**
 * Global variable for auto-setting of Regions flag.
 */
extern int KOLISEO_AUTOSET_REGIONS;
/**
 * Global variable for auto-setting of Koliseo_Temp Regions flag.
 */
extern int KOLISEO_AUTOSET_TEMP_REGIONS;

/**
 * Global variable for debug file pointer.
 */
extern FILE* KOLISEO_DEBUG_FP;

static const int KOLISEO_API_VERSION_INT = (KLS_MAJOR*1000000+KLS_MINOR*10000+KLS_PATCH*100); /**< Represents current version with numeric format.*/
static const char KOLISEO_API_VERSION_STRING[] = "0.2.0"; /**< Represents current version with MAJOR.MINOR.PATCH format.*/

const char* string_koliseo_version(void);

const int int_koliseo_version(void);

#define KLS_TITLEROWS 33
extern char* kls_title[KLS_TITLEROWS+1];

void kls_print_title_2file(FILE* fp);
void kls_print_title(void);

void kls_log(const char* tag, const char* format, ...);

#define KLS_DEFAULT_SIZE (16*1024) /**< Represents a simple default size for demo purposes.*/

#ifndef KLS_DEFAULT_ALIGNMENT
#define KLS_DEFAULT_ALIGNMENT (2*sizeof(void *)) /**< Represents a default alignment value. Not used.*/
#endif

/**
 * Represents a type index for Regions.
 * @see KLS_PUSH_TYPED()
 */
typedef enum Region_Type {
	None=0,
	Temp_KLS_Header=1,
	KLS_Header=2,
} Region_Type;

/**
 * Defines max index for Koliseo's own Region_Type values.
 * @see Region_Type
 */
#define KLS_REGIONTYPE_MAX KLS_Header

/**
 * Represents an allocated Region in a Koliseo.
 * @see KLS_PUSH()
 * @see KLS_PUSH_NAMED()
 */
typedef struct Region {
	ptrdiff_t begin_offset; /**< Starting offset of memory region.*/
	ptrdiff_t end_offset; /**< Ending offset of memory region.*/
	ptrdiff_t size; /**< Size of memory for the Region.*/
	char name[255]; /**< Name field for the Region.*/
	char desc[255]; /**< Description field for the Region.*/
	int type; /**< Used to identify which type the Region holds.*/
} Region;

static const char KOLISEO_DEFAULT_REGION_NAME[] = "No Name"; /**< Represents default Region name, used for kls_push_zero().*/
static const char KOLISEO_DEFAULT_REGION_DESC[] = "No Desc"; /**< Represents default Region desc, used for kls_push_zero().*/


#ifndef KOLISEO_LIST_H_
#define KOLISEO_LIST_H_
#include "stdbool.h"

typedef Region* element;

typedef struct list_region
{
	element value;
	struct list_region *next;
} region_list_item;
typedef region_list_item *Region_List;

/**
 * Represents the initialised arena allocator struct.
 * @see kls_new()
 * @see kls_clear()
 * @see kls_free()
 * @see KLS_PUSH()
 * @see KLS_POP()
 */
typedef struct Koliseo {
	char* data; /**< Points to data field.*/
	ptrdiff_t size; /**< Size of data field.*/
	ptrdiff_t offset; /**< Current position of memory pointer.*/
	ptrdiff_t prev_offset; /**< Previous position of memory pointer.*/
	Region_List regs; /**< List of allocated Regions*/
	int has_temp; /**< When == 1, a Koliseo_Temp is currently active on this Koliseo.*/
} Koliseo;

/**
 * Represents a savestate for a Koliseo.
 * @see kls_temp_start()
 * @see kls_temp_end()
 * @see KLS_PUSH()
 * @see KLS_POP()
 */
typedef struct Koliseo_Temp {
	Koliseo* kls; /**< Reference to the actual Koliseo we're saving.*/
	ptrdiff_t offset; /**< Current position of memory pointer.*/
	ptrdiff_t prev_offset; /**< Previous position of memory pointer.*/
	Region_List t_regs; /**< List of temporarily allocated Regions*/
} Koliseo_Temp;

ptrdiff_t kls_get_pos(Koliseo* kls);

Koliseo* kls_new(ptrdiff_t size);

//void* kls_push(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count);
void* kls_push_zero(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count);
void* kls_push_zero_AR(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count);
void* kls_push_zero_named(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count, char* name, char* desc);
void* kls_push_zero_typed(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count, int type, char* name, char* desc);
void* kls_pop(Koliseo* kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count);

#define KLS_PUSH(kls, type, count) (type*)kls_push_zero_AR(kls, sizeof(type), _Alignof(type), count)
#define KLS_PUSH_NAMED(kls, type, count, name, desc) (type*)kls_push_zero_named(kls, sizeof(type), _Alignof(type), count, name, desc)
#define KLS_PUSH_TYPED(kls, type, count, region_type, name, desc) (type*)kls_push_zero_typed(kls, sizeof(type), _Alignof(type), count, region_type, name, desc)
#define KLS_POP(kls, type, count) (type*)kls_pop(kls, sizeof(type), _Alignof(type), count)

#define KLS_PUSH_ARRAY(kls, type, count) (type*)kls_push_zero(kls, sizeof(type)*(count), _Alignof(type), count)
#define KLS_PUSH_STRUCT(kls, type) (type*)KLS_PUSH_ARRAY((kls), (type))

void kls_clear(Koliseo* kls);
void kls_free(Koliseo* kls);
void print_kls_2file(FILE* fp, Koliseo* kls);
void print_dbg_kls(Koliseo* kls);
void kls_formatSize(ptrdiff_t size, char* outputBuffer, size_t bufferSize);

#ifdef KOLISEO_HAS_CURSES

#ifndef KOLISEO_CURSES_H_
#define KOLISEO_CURSES_H_

#ifndef MINGW32_BUILD
#include "ncurses.h"
#else
#include <ncursesw/ncurses.h>
#endif

void kls_show_toWin(Koliseo* kls, WINDOW* win);
void kls_showList_toWin(Koliseo* kls, WINDOW* win);
void kls_temp_showList_toWin(Koliseo_Temp* t_kls, WINDOW* win);
#endif

#endif

Koliseo_Temp kls_temp_start(Koliseo* kls);
void kls_temp_end(Koliseo_Temp tmp_kls);
void* kls_temp_push_zero_AR(Koliseo_Temp t_kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count);
void* kls_temp_push_zero_named(Koliseo_Temp t_kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count, char* name, char* desc);
void* kls_temp_push_zero_typed(Koliseo_Temp t_kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count, int type, char* name, char* desc);
void* kls_temp_pop(Koliseo_Temp t_kls, ptrdiff_t size, ptrdiff_t align, ptrdiff_t count);

#define KLS_PUSH_T(kls_temp, type, count) (type*)kls_temp_push_zero_AR(kls_temp, sizeof(type), _Alignof(type), count)
#define KLS_PUSH_T_NAMED(kls_temp, type, count, name, desc) (type*)kls_temp_push_zero_named(kls_temp, sizeof(type), _Alignof(type), count, name, desc)
#define KLS_PUSH_T_TYPED(kls_temp, type, count, region_type, name, desc) (type*)kls_temp_push_zero_typed(kls_temp, sizeof(type), _Alignof(type), count, region_type, name, desc)
#define KLS_POP_T(kls_temp, type, count) (type*)kls_temp_pop(kls_temp, sizeof(type), _Alignof(type), count)

Region_List kls_emptyList(void);
#define KLS_GETLIST() kls_emptyList()
bool kls_empty(Region_List);
element kls_head(Region_List);
Region_List kls_tail(Region_List);
Region_List kls_cons(element, Region_List);

void kls_freeList(Region_List);
#define KLS_FREELIST(kls_list) kls_freeList(kls_list)
void kls_showList(Region_List);
void kls_showList_toFile(Region_List, FILE* fp);
#define KLS_ECHOLIST(kls_list) kls_showList(kls_list)
#define KLS_PRINTLIST(kls_list,file) kls_showList_toFile(kls_list,file)
bool kls_member(element, Region_List);
int kls_length(Region_List);
Region_List kls_append(Region_List, Region_List);
Region_List kls_reverse(Region_List);
Region_List kls_copy(Region_List);
Region_List kls_delete(element, Region_List);

Region_List kls_insord(element, Region_List);
#define KLS_PUSHLIST(reg,kls_list) kls_insord(reg,kls_list)
Region_List kls_insord_p(element, Region_List);
#define KLS_PUSHLIST_P(reg,kls_list) kls_insord_p(reg,kls_list)
Region_List kls_mergeList(Region_List, Region_List);
Region_List kls_intersect(Region_List, Region_List);
Region_List kls_diff(Region_List, Region_List);

#define KLS_DIFF(kls_list1,kls_list2) kls_diff(kls_list1,kls_list2)
bool kls_isLess(element, element);
bool kls_isEqual(element, element);
double kls_usageShare(element, Koliseo*);
void kls_usageReport_toFile(Koliseo*,FILE*);
void kls_usageReport(Koliseo*);
ptrdiff_t kls_type_usage(int, Koliseo*);

#endif

#endif
