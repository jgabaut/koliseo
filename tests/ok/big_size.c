// jgabaut @ github.com/jgabaut
// SPDX-License-Identifier: GPL-3.0-only
#include "../../src/koliseo.h"

int main(void) {
    Koliseo* k = kls_new(1024*1024*1024);

    int* p = KLS_PUSH(k, p);

    kls_free(k);
    printf("That's a big Koliseo.\n");
    return 0;
}
