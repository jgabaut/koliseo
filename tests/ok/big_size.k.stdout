That's a big Koliseo.$
